Issues with webdav backend

Hello,

Thanks for the great work. I was trying juicefs with webdav as storage backend but encountered issues. I used http://ajaxfilebrowser.com for testing, which provides a temporary webdav server.

1. Running the following command with the latest release `version 0.17.0 (2021-09-24T04:17:26Z e115dc4)`,
```
juicefs -v format  --storage webdav --bucket http://ajaxfilebrowser.com/User2fdbb37/  sqlite3://webdavjfs.sqlite3 webdavjfs
```
I got
```
2021/10/17 23:50:23.954912 juicefs[23236] <INFO>: Meta address: sqlite3://webdavjfs.sqlite3
[xorm] [info]  2021/10/17 23:50:23.955122 PING DATABASE sqlite3
2021/10/17 23:50:23.956194 juicefs[23236] <DEBUG>: Creating webdav storage at endpoint http://ajaxfilebrowser.com/User2fdbb37/
2021/10/17 23:50:23.956230 juicefs[23236] <INFO>: Data use webdav://ajaxfilebrowser.com/webdavjfs/
panic: interface conversion: interface {} is *bytes.Buffer, not *[]uint8

goroutine 1 [running]:
github.com/juicedata/juicefs/pkg/object.(*webdav).Put(0xc00099b990, 0xc0008c2800, 0x1c, 0x6e064c0, 0xc00016b440, 0x0, 0x0)
        /go/src/github.com/juicedata/juicefs/pkg/object/webdav.go:109 +0x565
github.com/juicedata/juicefs/pkg/object.(*withPrefix).Put(0xc0008bd9e0, 0xc0008c22a0, 0x12, 0x6e064c0, 0xc00016b440, 0x6e07040, 0xc0008bddc0)
        /go/src/github.com/juicedata/juicefs/pkg/object/prefix.go:61 +0xa6
main.doTesting(0x6e8b160, 0xc0008bd9e0, 0xc0008c22a0, 0x12, 0xc0002a82a0, 0x64, 0x64, 0x6a16cc8, 0xc0011a1840)
        /go/src/github.com/juicedata/juicefs/cmd/format.go:100 +0x6cf
main.test(0x6e8b160, 0xc0008bd9e0, 0x0, 0x0)
        /go/src/github.com/juicedata/juicefs/cmd/format.go:132 +0x1dd
main.format(0xc000990c80, 0xd, 0x18)
        /go/src/github.com/juicedata/juicefs/cmd/format.go:208 +0xc92
github.com/urfave/cli/v2.(*Command).Run(0xc000820120, 0xc0009908c0, 0x0, 0x0)
        /go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/command.go:163 +0x4ed
github.com/urfave/cli/v2.(*App).RunContext(0xc0008009c0, 0x6e605c0, 0xc00019e008, 0xc0005ace00, 0x9, 0x10, 0x0, 0x0)
        /go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:313 +0x81f
github.com/urfave/cli/v2.(*App).Run(...)
        /go/pkg/mod/github.com/urfave/cli/v2@v2.3.0/app.go:224
main.main()
        /go/src/github.com/juicedata/juicefs/cmd/main.go:100 +0x1b6c
```
This can be fixed by reverting https://github.com/juicedata/juicefs/pull/800.

2. After fixing that error, rerun the command and I got

```
2021/10/18 00:19:57.628568 juicefs[35628] <INFO>: Meta address: sqlite3://webdavjfs.sqlite3
[xorm] [info]  2021/10/18 00:19:57.628813 PING DATABASE sqlite3
2021/10/18 00:19:57.629894 juicefs[35628] <WARNING>: The latency to database is too high: 1.097259ms
2021/10/18 00:19:57.630117 juicefs[35628] <DEBUG>: Creating webdav storage at endpoint http://ajaxfilebrowser.com/User2fdbb37/
2021/10/18 00:19:57.630146 juicefs[35628] <INFO>: Data use webdav://ajaxfilebrowser.com/webdavjfs/

2021/10/18 00:20:10.558519 juicefs[35628] <FATAL>: Storage webdav://ajaxfilebrowser.com/webdavjfs/ is not configured correctly: Failed to put: 409 Conflict: <error xmlns="DAV:"><responsedescription xmlns="DAV:">Item&#39;s parent doesn&#39;t exist</responsedescription></error>
```

I dig into this and turns out that the function https://github.com/juicedata/juicefs/blob/main/pkg/object/webdav.go#L98-L100 is not able to catch all the possible cases.
I managed to fix the issue with the following very hacky patch:

```diff
diff --git a/pkg/object/webdav.go b/pkg/object/webdav.go
index 1105ec57..c0b6623f 100644
--- a/pkg/object/webdav.go
+++ b/pkg/object/webdav.go
@@ -96,7 +96,7 @@ func (w *webdav) mkdirs(p string) error {
 }

 func (w *webdav) isNotExist(err error) bool {
-       return strings.Contains(err.Error(), "Not Found")
+       return strings.Contains(err.Error(), "Not Found") || strings.Contains(err.Error(), "Item&#39;s parent doesn&#39;t exist") || strings.Contains(err.Error(), "Parent node does not exist")
 }

 func (w *webdav) Put(key string, in io.Reader) error {
```

The two additional conditions are based on what I see from webdav server of http://ajaxfilebrowser.com and Box Drive.

After fixing these, I managed to run juicefs on webdav. However my solution to 2 is obviously a bad one, and I am not sure what would be the correct way.

--- a/pkg/object/webdav.go
+++ b/pkg/object/webdav.go
@@ -87,7 +87,7 @@ func (w *webdav) Get(key string, off, limit int64) (io.ReadCloser, error) {
 
 func (w *webdav) mkdirs(p string) error {
 	err := w.c.Mkdir(p)
-	if err != nil && w.isNotExist(p) {
+	if err != nil && w.isNotExist(path.Dir(p)) {
 		if w.mkdirs(path.Dir(p)) == nil {
 			err = w.c.Mkdir(p)
 		}

